numpy/core/src/umath/string_ufuncs.cpp

```diff
@@ -0,0 +1,448 @@
+    spec.nout = 1;
+    spec.dtypes = dtypes;
+    spec.slots = slots;
+    spec.flags = NPY_METH_NO_FLOATINGPOINT_ERRORS;
+
+    /* Use this loop variable for typing more explicitly */
+    PyArrayMethod_StridedLoop *loop;
+
+    /* TODO: It would be nice to condense the below */
+    /* All String loops */
+    loop = string_comparison_loop<false, Py_EQ, npy_byte>;
+    if (add_loop(umath, "equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_NE, npy_byte>;
+    if (add_loop(umath, "not_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_LT, npy_byte>;
+    if (add_loop(umath, "less", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_LE, npy_byte>;
+    if (add_loop(umath, "less_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_GT, npy_byte>;
+    if (add_loop(umath, "greater", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_GE, npy_byte>;
+    if (add_loop(umath, "greater_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+
+    /* All Unicode loops */
+    dtypes[0] = Unicode;
+    dtypes[1] = Unicode;
+
+    loop = string_comparison_loop<false, Py_EQ, npy_ucs4>;
+    if (add_loop(umath, "equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_NE, npy_ucs4>;
+    if (add_loop(umath, "not_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_LT, npy_ucs4>;
+    if (add_loop(umath, "less", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_LE, npy_ucs4>;
+    if (add_loop(umath, "less_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_GT, npy_ucs4>;
+    if (add_loop(umath, "greater", &spec, loop) < 0) {
+        goto finish;
+    }
+    loop = string_comparison_loop<false, Py_GE, npy_ucs4>;
+    if (add_loop(umath, "greater_equal", &spec, loop) < 0) {
+        goto finish;
+    }
+
+    res = 0;
+  finish:
+    Py_DECREF(String);
+    Py_DECREF(Unicode);
+    Py_DECREF(Bool);
+    return res;
+}
+
+
+template <bool rstrip, typename character>
+static PyArrayMethod_StridedLoop *
+get_strided_loop(int comp)
+{
+    switch (comp) {
+        case Py_EQ:
+            return string_comparison_loop<rstrip, Py_EQ, character>;
+        case Py_NE:
+            return string_comparison_loop<rstrip, Py_NE, character>;
+        case Py_LT:
+            return string_comparison_loop<rstrip, Py_LT, character>;
+        case Py_LE:
+            return string_comparison_loop<rstrip, Py_LE, character>;
+        case Py_GT:
+            return string_comparison_loop<rstrip, Py_GT, character>;
+        case Py_GE:
+            return string_comparison_loop<rstrip, Py_GE, character>;
+        default:
+            assert(false);
```

Comment:
ok, seeing the same assert again, what you probably should do is

```
static constexpr is_valid_comp(int comp) {
   return comp == Py_LT || ...;
}
```
and use it either in fonction prelude, either as an `assert` or a `static_assert` depending on the context

Or even better: make the int an `enum class` but that's probably not possible, just mentioning it.

Reply:
OK, I didn't love the idea of a long duplicate `Py_EQ || ...` where the asserts were, but if it goes into the function prelude, I can see the point, will adopt :).

Reply:
clang static analysis seems to not be quite smart enough to realize that the switch will always match, but I guess I will ignore that (it warns that `res` may be undefined in the other case).

Reply:
I guess I could use `default` for the last one, but dunno.

Reply:
Oh, silly me... This assert is actually not static at all.

Reply:
Sorry for the verboseness. Pushed a change with that the enum class instead, figured that the static assert was just as verbose anyway...